api/server.go

```diff
@@ -33,5 +40,36 @@ type Server struct {
 }
 
 func (s *Server) Stop() {
-	helper.Logger.Info("Server stopped")
+	stopping = true
+	helper.Logger.Info("Stopping API server")
+	for {
+		n := atomic.LoadInt64(&runningRequests)
+		if n == 0 {
+			break
+		}
+		time.Sleep(time.Second)
+	}
+	helper.Logger.Info("API Server stopped")
+}
+
+// FIXME this is ugly
+var stopping = false
+var runningRequests int64 = 0
+
+type gracefulHandler struct {
+	handler http.Handler
+}
+
+func (l gracefulHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if stopping {
+		WriteErrorResponse(w, r, ErrMaintenance)
+		return
+	}
+	atomic.AddInt64(&runningRequests, 1)
+	defer atomic.AddInt64(&runningRequests, -1)
```

Comment:
If an exception error occurs in a request (such as a null pointer exception, of course it may be the cause of the code), will the defer statement never be executed?

Reply:
The execution is ensured, that's why we use `defer`